ccdproc/core.py

```diff
@@ -1558,6 +1558,144 @@ def cosmicray_median(ccd, error_image=None, thresh=5, mbox=11, gbox=0,
+    scaled by the square root of the number of pixels summed so that
+    statistically low or high column regions may be detected even though
+    individual pixels may not be statistically deviant. For the purpose of this
+    task one would normally select large sigma threshold factors such as six or
+    greater to detect only true bad pixels and not the extremes of the noise
+    distribution.
+
+    As a final step each column is examined to see if there are small segments
+    of unflagged pixels between bad pixels. If the length of a segment is less
+    than that given by the ngood parameter all the pixels in the segment are
+    also marked as bad.
+
+
+    Returns
+    -------
+    mask : `numpy.ndarray`
+        A boolean ndarray with the bad pixel positions identified.
+    '''
+    if ratio.data.ndim !=2:
+        return None
+    mask = ~np.isfinite(ratio.data)
+    medsub = ratio.data - ndimage.filters.median_filter(ratio.data,
+                                                        size=(nlsig, ncsig))
+
+    nlines, ncols = ratio.data.shape
+    if byblocks is True:
+        nlinesblock = int(np.ceil(nlines / nlsig))
+        ncolsblock = int(np.ceil(ncols / ncsig))
+        for i in range(nlinesblock):
+            for j in range(ncolsblock):
+                l1 = i*nlsig
+                l2 = min((i+1)*nlsig, nlines)
+                c1 = j*ncsig
+                c2 = min((j+1)*ncsig, ncols)
+                block = medsub[l1:l2,c1:c2]
+                high = np.percentile(block.ravel(), 69.1)
+                low = np.percentile(block.ravel(), 30.9)
+                block_sigma = (high-low)/2.0
+                block_mask = ( (block > hsigma*block_sigma) |
+                               (block < -lsigma*block_sigma) )
+                mblock = np.ma.MaskedArray(block, mask=block_mask)
+
+                if findbadcolumns:
+                    csum = np.ma.sum(mblock, axis=0)
+                    csum[csum<=0] = 0
+                    csum_sigma = np.ma.MaskedArray(np.sqrt(c2-c1-csum))
+                    colmask = ( (csum.filled(1) > hsigma*csum_sigma) |
+                                (csum.filled(1) < -lsigma*csum_sigma) )
+                    for c in range(c2-c1):
+                        block_mask[:,c] = block_mask[:,c] | np.array([colmask[c]]*(l2-l1))
+
+                mask[l1:l2,c1:c2] = block_mask
+    else:
+        high = ndimage.percentile_filter(medsub, 69.1, size=(nlsig, ncsig))
+        low = ndimage.percentile_filter(medsub, 30.9, size=(nlsig, ncsig))
+        sigmas = (high-low)/2.0
+        mask = ( (medsub > hsigma*sigmas) | (medsub < -lsigma*sigmas) )
+
+    if findbadcolumns:
+        # Loop through columns and look for short segments (<ngood pixels long)
+        # which are unmasked, but are surrounded by masked pixels and mask them
+        # under the assumption that the column region is bad.
+        for col in range(0,ncols):
```

Comment:
could you use six.moves.range here and in the following lines to avoid python2 range?